### DISCUSSION TITLE
How to handle useQuery running on the server

### DISCUSSION THREAD

**Discussion by seonghyeonkimm** [TanStack/query#3030] [Q&A] [answered]
### Descriptions
- I am using react-query with suspense enabled in my nextjs project.
- When nextjs trying to render my app on the server, it trigger react-query useQuery to be run. And the useQuery throw error because it can't be succeed on the server. I only want my useQuery to be run on the client. it is making my log messages messy.
- if there is anybody who handle this kind of errors well, please share the ideas on this problem.

for now, I am handling this problem like below.

```
function useMyQuery(queryKey, queryFn) {
  return useQuery(queryKey, async () => {
    if (typeof window === 'undefined') {
      return {};
    }

    const result = await queryFn();
    return result;
  }
}
```

**Comment by TkDodo** [comment, marked as answer]
if you run your code on the server, you have to make sure that it works on the server as well. From a `useQuery` perspective, I don't think there's anything that doesn't work on the server.

**Reply by seonghyeonkimm** [reply]
Thanks for your answer. However, I don't think that running query on the server is not necessary cause it won't be used for creating html on the server first of all. And it could be wasting api server resource. What do you think about this kind of problem?

**Reply by TkDodo** [reply]
can you create a codesandbox example please? because the react-query example uses prefetching and I don't see the queryFn being called on the server in that setup 🤔

**Reply by seonghyeonkimm** [reply]
Okay. Let me create the sample to show you what problem I am having exactly.

To share first breifly before the sample codesandbox, in my case, I am talking about useQuery which is not prefetched on the server.

**Reply by TkDodo** [reply]
I think I found it out. `useQuery` doesn't fire a fetch on the server, because we trigger a subscription + the fetch on mount in a useEffect - which never runs on the server 😅 . This is why you need `prefetching` and `hydration` to pre-render your data.

**Reply by seonghyeonkimm** [reply]
Oh, really? In my case, I saw the error logs related to api request on my ecs logs. So, I thought it is triggered on the server. Let me try to figure out more details and come back to this issue. Thanks.

**Reply by seonghyeonkimm** [reply]
Sorry for coming back late. I just reproduced the condition where useQuery runs on the server. And it is also the same environment with my workspace.

1. Use nextjs for react SSR framework.
2. Update react@next, react-dom@next.
3. Enable QueryClient suspense option.

Then, useQuery runs on the server. too. I pushed the example codes on [the repository](https://github.com/seonghyeonkimm/next-react-query). Do you know why useQuery runs on the server in this condition? I can guess that it is related to suspense options only.

**Reply by TkDodo** [reply]
with suspense, we trigger data fetching during render and throw the promise, because that's what suspense requires. I guess this also runs on the server. but why would you need `react@next` to reproduce it - wouldn't that be the same in v17?

@Ephem FYI, maybe you know if suspense queries _should_ run on the server, too?

**Reply by seonghyeonkimm** [reply]
Ah ha..! As you mentioned, it is how suspense works. Thanks for your reply. The reason why I use Suspense with `react@next`, `react-dom@next` is that nextjs throw errors like `Error: ReactDOMServer does not yet support Suspense` when using react v17.

Now, I understand why useQuery runs on the server too when suspense option is enabled. However, in my case useQuery on the server can't succeed because it can't get access to cookies. And also it will be prefetched on the serve and will be rendered without fetching data again. What approcah do you think is appropriate in my case? if you can come up with something, Could I ask you a opinion on this issue?

**Reply by Ephem** [reply]
Suspense does not work on the server in React 17, but Suspense itself will work in React 18.

Data fetching with Suspense will not officially be supported by the React core team in the initial React 18 release, so it's unclear to what extent we will be able to support Suspense on the server in RQ initially.

@seonghyeonkimm The normal way of doing SSR with RQ is documented in [this guide](https://react-query.tanstack.com/guides/ssr). It does not use Suspense, instead you prefetch the queries you need in `getStaticProps` or `getServerSideProps` so the data is available before the components even render.

**Reply by Ephem** [reply]
If you want queries with Suspense to only run on the client, you could try setting `enabled: typeof window !== 'undefined'`, but that _might_ give you hydration warnings. A sure way to do it would be something like:

```
function useShouldLoad() {
  const [shouldLoad, setShouldLoad] = useState(false);
  useEffect(() => {
    setShouldLoad(true);
  }, []);
  return shouldLoad;
}

function Comp() {
  const shouldLoad = useShouldLoad();
  const result = useQuery(queryKey, fetchFn, { enabled: shouldLoad });
  
  // ...
}
```

This would enable the query _after_ hydration, so you know the query will only run on the client and hydration will match what was rendered on the server.

**Reply by Ephem** [reply]
Oh, and the use case you are describing "In React 18 with SSR, I want to run Suspense enabled queries only on the client" might be a common enough one that it's something we should discuss supporting out of the box.

**Reply by seonghyeonkimm** [reply]
Thanks for your reply. I realized that It happened because I am using react, react-query in the way that these libraries are not supporting officially now. Then, I think only way I could use is just adding if statement to queryFunction in my case for time being. maybe until react and react-query supports this case officially.

Thanks for your kind guidance again, both of you. @TkDodo @Ephem

**Reply by seonghyeonkimm** [reply]
Yes. ["In React 18 with SSR, I want to run Suspense enabled queries only on the client"](https://github.com/tannerlinsley/react-query/discussions/3030#discussioncomment-1786433) is exactly what I want.

And, as you mentioned above, I tried enabled options too before using if statement. But I could not use suspense mode with enabled options.

**Reply by TkDodo** [reply]
> Oh, and the use case you are describing "In React 18 with SSR, I want to run Suspense enabled queries only on the client" might be a common enough one that it's something we should discuss supporting out of the box.

agreed, lets try to make this happen once we know more about react ssr in 18 :)
